[#47989] Create Boards outside of a Project

aaron-contreras opened the pull request:
Most of the heavy-lifting is set up on the new `Boards::BaseCreateService`. Subclasses of it can rely on its defaults and implement their own "custom" behavior such as `Boards::VersionBoardCreateService` which will create a `Grids::Widget` and `Query` per version.

**NOTE**: I ripped off the sass file from the Angular new board modals. There were a couple of things that needed adapting with the new layout specified in the Figma file. If there's cleaner/better way to accomplish this, do tell 😄!

See: https://community.openproject.org/work_packages/47989

aaron-contreras committed:
Add `boards/new` routing

* Sets up the `/boards/new` route to the `/boards/boards#new` action
* Adds a routing spec to verify proper routing
* Replace `{ }` multi-line block syntax for `do; end`

aaron-contreras committed:
Add "Create New Board" button to Overview Page

Extends the `AddButtonComponent` View Component and sets up specs
to demonstrate desired button rendering behavior.

**NOTE**: Small refactor of spec setup to improve clarity and performance.

aaron-contreras committed:
Add `POST /boards` routing

Adds a route in order for the global create form to submit to.

aaron-contreras committed:
Add ability to create "Basic" Boards globally

aaron-contreras committed:
Add ability to create "Status" boards globally

aaron-contreras committed:
Add ability to create "Assignee" boards globally

aaron-contreras committed:
Add ability to create "Version" boards globally

* Refactors the single `GlobalCreateService` into individual services
that implement a `BaseCreateService` with shared/common functionality.

aaron-contreras committed:
Add ability to create "Subproject" boards globally

aaron-contreras committed:
Add ability to create "Parent-child" boards globally

aaron-contreras committed:
Add specs for validation errors

* Refactors the `instance` method for the `BaseCreateService` to be
less generalized given we're specifically working with `Boards::Grid`
objects in this service.

aaron-contreras committed:
Style radio buttons as tiles

aaron-contreras committed:
Select first radio button by default

aaron-contreras committed:
Refactor `grid_lacks_query?` and remove unused parameter lists

aaron-contreras committed:
Set Widget Queries to be public

aaron-contreras committed:
Add spec for `Boards::CreateContract`

aaron-contreras committed:
Add spec for `Boards::AssigneeBoardCreateService`

aaron-contreras committed:
Add spec for `Boards::VersionBoardCreateService`

aaron-contreras committed:
Add spec for `Boards::StatusBoardCreateService`

aaron-contreras committed:
Add spec for `Boards::BasicBoardCreateService`

aaron-contreras committed:
Add spec for `Boards::SubprojectBoardCreateService`

aaron-contreras committed:
Add spec for `Boards::SubtasksBoardCreateService`

aaron-contreras committed:
Appease my buddy Rubocop!

oliverguenther commented:
- [x] When in community edition, only the basic boards can be selected (Right now you can create them but are redirected to upsale page)

oliverguenther commented:
The queries created currently do not include the sortBy criteria "manualSorting-asc" which breaks the manual sorting / drag&drop sorting right now.

The sort criteria needs to be `[[:manual_sorting, 'asc'], [:id, 'asc']]` to match this: [frontend/src/app/features/boards/board/board-list/board-lists.service.ts](https://github.com/opf/openproject/blob/17b47504d364f65fb87e9abc43587b8ae487bc31/frontend/src/app/features/boards/board/board-list/board-lists.service.ts#L107-L110)

aaron-contreras committed:
Add Enterprise Edition restriction to "action" board types

aaron-contreras committed:
Add default `sort_criteria` to queries created for Boards

aaron-contreras committed:
Merge branch 'dev' into implementation/47989-create-boards-outside-of-a-project

oliverguenther commented on modules/boards/app/controllers/boards/boards_controller.rb:
```diff
@@ -37,6 +37,8 @@ def overview
     def new; end
 
     def create
+      return render_403 if restricted_board_type?
```
I'd prefer to see this as a before action to prevent the extra return, but that's just my preference

aaron-contreras replied:
Addressed in https://github.com/opf/openproject/pull/13259/commits/f82d2e3588cbcb599daf85bccfec65b317d90875

aaron-contreras committed:
Extract restricted boards check into a `before_action` hook

oliverguenther commented on frontend/src/global_styles/content/_tiles.sass:
```diff
@@ -0,0 +1,92 @@
+
+.form--field
+  &.-align-start
+    align-items: start
```
We usually put overrides like that into its separate file, to retain the bem block as the only content of it. You could add it to either `forms.sass` or add something like `_boards_overrides.sass` in another page-specific file. It does not belong to _tiles

oliverguenther commented on modules/boards/app/controllers/boards/boards_controller.rb:
```diff
@@ -1,6 +1,9 @@
 module ::Boards
   class BoardsController < BaseController
+    include ::BoardsHelper
```
If I'm not confusing something, you don't need to include the helper, you could just use `helpers.<method>`?

aaron-contreras replied:
You're correct.

As I just wrote in the commit message to remove it:

> I was mislead into this being needed during my testing as `BoardsHelper`
> didn't seem to be available within the view but it seemed to be because
> there was no `helpers` directory initially so it wasn't going to be autoloaded
> until I stopped spring and restarted the rails server.

oliverguenther reviewed (approved):
Some tiny remarks left @aaron-contreras , but overall excellent job! Feel free to merge once you addressed / commented the points.

aaron-contreras committed:
Move form styles out of `_tiles.sass`

As suggested, these styles better belong in `_forms.sass`

aaron-contreras committed:
Remove unnecessary `include` of `BoardsHelper`

I was mislead into this being needed during my testing as `BoardsHelper`
didn't seem to be available within the view but it seemed to be because
there was no `helpers` directory initially so it wasn't going to be autoloaded
until I stopped spring and restarted the rails server.

This is indeed unnecessary; Thanks Oliver.
